add a new blog post introducing MIR

nikomatsakis opened the pull request:
The images don't work in this local version because of the absolute paths. You can view it here: 

http://nikomatsakis.github.io/blog.rust-lang.org/2016/04/19/MIR.html

cc @rust-lang/core @rust-lang/compiler

rust-highfive commented:
r? @aturon

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* Pull requests are usually filed against the master branch for this repo, but this one is against gh-pages. Please double check that you specified the right target!

ubsan commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+
+That is, we are introducing a new intermediate representation (IR) of
+your program that we call **MIR**: MIR stands for *middle* IR, because
+the MIR comes between the existing HIR ("high-level IR", roughly an
```
Nit: mid-level IR, not middle IR.

michaelwoerister replied:
I agree.

tsion commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+  with MIR is just plain *easier*, because it contains a much more
+  primitive set of operations than ordinary Rust. This simplification
+  enables us to do a lot of things that were forbodingly complex
+  before. We'll look at one such case in those post -- non-zeroing
```
s/those post/this post/

tsion commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+
+The idea of "simplifying" a loop into uses of `goto` (in the form of a
+CFG) is sort of an interesting one. `goto` is in fact a *more general*
+construct that `loop`, and it's something we would not want to
```
s/that `loop`/than `loop`/

tsion commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+which we extract the various elements. Well, this value `iterator`
+actually has a destructor, and it will need to be freed (in this case,
+its job is to free the memory that was used by the vector `vec`; this
+memory is no longer needed, since we've finished iterator over the
```
s/iterator/iterating/

tsion commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+roughly the same thing that the equivalent C++ code would do.)
+
+But we'd like to do better than that. What we would *like* to do is to
+using boolean flags on the stack that tell us what needs to be freed.
```
s/using/use/

nrc commented:
Nice post! I guess the audience is 'average Rust user'. I think if that is the case, then it could be improved slightly by being more concrete about the user-facing goals (e.g., give an example of code which currently errors, but would succeed with NLLs) and explain some of the jargon inline rather than give links.

With a few extras, I think the blog could be really interesting to an audience of 'people with an interest in compilers'. For that audience, I think the post would benefit from:

* more discussion of how MIR fits into type checking, i.e., it relies on type checking, is used for borrow checking, etc. (c.f., the first flow chart, stating why we need type info to construct the MIR as opposed to just saying where it fits into the process)
* comparison with SSA (and other options) why MIR isn't SSA and why you don't use it
* what isn't explicit in the MIR (thinking of fat pointers for example, but also other trans stuff like DPS)
* what temporaries exist in the MIR (lvalues vs rvalues vs allocas (in trans))

I think that you could add this stuff without making the article much longer and make it really interesting to a bunch more people.

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+    fertile ground for going beyond LLVM's optimizations.
+
+    In addition, MIR will uncork some longstanding performance improvements
+    to the code Rust generates, like ["non-zeroing" drop][rfc320].
```
Given that we dive into non-zeroing drop later on, it would be nice if you could explain what it is up front (around here), rather than just linking.

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+commit. Once we're able to run [Crater][crater] with MIR enabled and
+see no regressions across the entire [crates.io](http://crates.io/)
+ecosystem, we'll turn it on by default (or, you'll forgive a terrible
+(wonderful) pun, [launch MIR into orbit][orbit]).
```
I would cut this para or replace it with something much shorter - "MIR is in the compiler, will be turned on soon-ish"

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+implementing [non-zeroing drops][rfc320], a long-desired
+optimization. If after this post you find you are hungry for more,
+have a look at the [RFC that introduced MIR][rfc1211], or jump right
+into [the code][].
```
missing link

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+
+By applying this rewriting, we can remove all `for` loops, but that
+still leaves multiple kinds of loops. So next we can imagine rewrite
+all `while let` loops into a simple `loop` combined with a `match`:
```
Maybe worth mentioning that this step is done in the AST-> HIR lowering (you already mentioned the HIR in the intro)

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+// Rather than `vec.into_iter()`, we are calling
+// the function `IntoIterator::into_iter`. This is
+// exactly equivalent, just more explicit.
+let mut iterator = IntoIterator::into_iter(vec);
```
I would have used UFCS syntax here, is there a reason not to?

eddyb replied:
It keeps the line shorter and doesn't have semantic implications, I think the point was method call desugaring, not path resolution.

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+the `loop` using that.
+
+Here is the running example from the preivous section, expressed as a
+control-flow graph:
```
I would like to see this as pseudo code with `goto`s and labels (just this one, not all the CFGs, and as well as the diagram). Maybe I'm just weird, but for me it is easier to make the transition from code to graph with that extra step.

eddyb replied:
I agree, I was surprised not to see goto or even just labelled break (to introduce goto).

nrc commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+such as to express a `loop` or a `break`.
+
+### Simplifying match expressions
+
```
I feel like somewhere before here you should mention that the MIR is typed.

huonw commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+representation effectively has a general `goto`, and so we can express
+the `loop` using that.
+
+Here is the running example from the preivous section, expressed as a
```
"preivous" -> "previous"

huonw commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+                      [ return ]
+-->
+
+We can then transform this graph by identifying each place where
```
This paragraph and the next image don't appear in the rendered version, it looks like the previous comment isn't ended correctly: the closing `-->` appears to be converted into &ndash;> (`&ndash;>`).

michaelwoerister commented on _posts/2016-04-19-MIR.md:
```diff
@@ -0,0 +1,725 @@
+though, MIR is based on the notion of a
+[control-flow graph (CFG)][CFG]. If you've ever used a flow-chart,
+then the concept of a control-flow graph will be pretty familiar to
+you. It's a representation of your program that exploses the
```
exploses => exposes?

nikomatsakis commented:
Updated to address various bits of feedback. I reorganized the flow somewhat, introduced `goto`, and tried to give a bit more detail that may interest compiler enthusiasts, but without being too overwhelming (I think the primary audience is going to be more general than that).
